aya/src/sys/bpf.rs

```diff
@@ -371,6 +441,233 @@ pub(crate) fn bpf_raw_tracepoint_open(name: Option<&CStr>, prog_fd: RawFd) -> Sy
     sys_bpf(bpf_cmd::BPF_RAW_TRACEPOINT_OPEN, &attr)
 }
 
-fn sys_bpf(cmd: bpf_cmd, attr: &bpf_attr) -> SysResult {
+pub(crate) fn bpf_load_btf(raw_btf: &[u8], log: &mut VerifierLog) -> Result<RawFd, io::Error> {
+    let mut attr = unsafe { mem::zeroed::<bpf_attr>() };
+    let u = unsafe { &mut attr.__bindgen_anon_7 };
+    u.btf = raw_btf.as_ptr() as *const _ as u64;
+    u.btf_size = mem::size_of_val(raw_btf) as u32;
+    let log_buf = log.buf();
+    if log_buf.capacity() > 0 {
+        u.btf_log_level = 1;
+        u.btf_log_buf = log_buf.as_mut_ptr() as u64;
+        u.btf_log_size = log_buf.capacity() as u32;
+    }
+    match sys_bpf(bpf_cmd::BPF_BTF_LOAD, &attr) {
+        Ok(v) => Ok(v as RawFd),
+        Err((_, err)) => Err(err),
+    }
+}
+
+pub(crate) fn bpf_btf_get_fd_by_id(id: u32) -> Result<RawFd, io::Error> {
+    let mut attr = unsafe { mem::zeroed::<bpf_attr>() };
+    attr.__bindgen_anon_6.__bindgen_anon_1.btf_id = id;
+
+    match sys_bpf(bpf_cmd::BPF_BTF_GET_FD_BY_ID, &attr) {
+        Ok(v) => Ok(v as RawFd),
+        Err((_, err)) => Err(err),
+    }
+}
+
+pub(crate) fn is_prog_name_supported() -> bool {
+    let mut attr = unsafe { mem::zeroed::<bpf_attr>() };
+    let u = unsafe { &mut attr.__bindgen_anon_3 };
+    let mut name: [c_char; 16] = [0; 16];
+    let cstring = CString::new("test").unwrap();
+    let name_bytes = cstring.to_bytes();
+    let len = min(name.len(), name_bytes.len());
+    name[..len].copy_from_slice(unsafe {
+        slice::from_raw_parts(name_bytes.as_ptr() as *const c_char, len)
+    });
+    u.prog_name = name;
+
+    let prog: &[u8] = &[
+        0xb7, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // mov64 r0, 0
```

Comment:
The comment should say `r0 = 0` which is BPF asm format

Reply:
> A few more comments.
> 
> How are you testing this? I'm 99% sure that setting funcinfo/lineinfo isn't working (because of that Vec stuff).

I have a test program - i'll push it tomorrow so you can have a play.

> Do we need to parse funcinfo/lineinfo at all? 

Yes. You can't load BTF to the kernel without it (I tried and failed).

> Can't we just proxy it verbatim to the kernel? (legit question, it's not clear to me at the moment)

No. Because the ELF and kernel API are subtly different in that offsets in ELF format are in bytes, whereas in kernel format they are in instructions. So we have to parse it (unfortunately).